src/irmin-pack/inode.ml

```diff
@@ -992,6 +1030,78 @@ struct
       check_lower t
 
     let is_tree t = match t.v with Tree _ -> true | Values _ -> false
+
+    type nonrec proof = (hash, step, value) Irmin.Private.Node.Proof.t
+    [@@deriving irmin]
+
+    module Proof = struct
+      let rec proof_of_concrete h : Concrete.t -> proof = function
+        | Blinded -> Blinded (Lazy.force h)
+        | Values vs -> Values (List.map Concrete.of_entry vs)
+        | Tree tr ->
+            let proofs =
+              List.fold_left
+                (fun acc (e : _ Concrete.pointer) ->
+                  let p = proof_of_concrete (lazy e.pointer) e.tree in
+                  let e = (e.index, p) in
+                  e :: acc)
+                [] (List.rev tr.pointers)
+            in
+            Inode { length = tr.length; proofs }
+
+      let hash_v v = Bin.V.hash (to_bin_v Truncated v)
+
+      let rec hash : int -> proof -> hash =
+       fun depth -> function
+        | Values l -> hash_v (Values (StepMap.of_list l))
+        | Inode { length; proofs } ->
+            let es =
+              List.fold_left
+                (fun acc (index, proof) ->
+                  let pointer = hash (depth + 1) proof in
+                  (index, Broken pointer) :: acc)
+                [] proofs
+            in
+            let entries = Array.make Conf.entries None in
+            List.iter (fun (index, ptr) -> entries.(index) <- Some ptr) es;
+            let v : truncated_ptr v = Tree { depth; length; entries } in
+            hash_v v
+        | Blinded h -> h
+
+      let rec concrete_of_proof depth : proof -> Concrete.t = function
+        | Blinded _ -> Concrete.Blinded
+        | Values vs -> Concrete.Values (List.map Concrete.to_entry vs)
+        | Inode { length; proofs } ->
+            let pointers =
+              List.fold_left
+                (fun acc (index, proof) ->
+                  let tree = concrete_of_proof (depth + 1) proof in
+                  let pointer = hash (depth + 1) proof in
+                  { Concrete.tree; pointer; index } :: acc)
+                [] (List.rev proofs)
+            in
+            Concrete.Tree { depth; length; pointers }
+
+      let to_proof la t =
+        let p =
+          if t.stable then
+            (* To preserve the stable hash, the proof needs to contain
+               all the underlying values. *)
+            let bindings =
+              seq la t |> Seq.map Concrete.to_entry |> List.of_seq
+            in
+            Concrete.Values bindings
+          else to_concrete ~force:false la t
+        in
+        proof_of_concrete t.hash p
+
+      let of_proof (proof : proof) =
+        let c = concrete_of_proof 0 proof in
+        of_concrete_exn c
+
+      let of_concrete t = proof_of_concrete (lazy (failwith "blinded root")) t
```

Comment:
This function doesn't check for stability as `to_proof` does, I think that this is a problem